Content.Client/Xenoarchaeology/Ui/XenoArtifactGraphControl.xaml.cs

```diff
@@ -0,0 +1,180 @@
+using System.Linq;
+using System.Numerics;
+using Content.Client.Xenoarchaeology.Artifact;
+using Content.Shared.Xenoarchaeology.Artifact.Components;
+using Robust.Client.AutoGenerated;
+using Robust.Client.Graphics;
+using Robust.Client.ResourceManagement;
+using Robust.Client.UserInterface;
+using Robust.Client.UserInterface.Controls;
+using Robust.Client.UserInterface.XAML;
+using Robust.Shared.Input;
+
+namespace Content.Client.Xenoarchaeology.Ui;
+
+[GenerateTypedNameReferences]
+public sealed partial class XenoArtifactGraphControl : BoxContainer
+{
+    [Dependency] private readonly IEntityManager _entityManager = default!;
+
+    private Entity<XenoArtifactComponent>? _artifact;
+
+    private Entity<XenoArtifactNodeComponent>? _hoveredNode;
+
+    private readonly Font _font;
+
+    public event Action<Entity<XenoArtifactNodeComponent>>? OnNodeSelected;
+
+    private static Color _lockedNodeColor = Color.FromHex("#777777");
+
+    private float NodeRadius => 25 * UIScale;
+    private float NodeDiameter => NodeRadius * 2;
+    private float MinYSpacing => NodeDiameter * 0.75f;
+    private float MaxYSpacing => NodeDiameter * 1.5f;
+    private float MinXSpacing => NodeDiameter * 0.33f;
+    private float MaxXSpacing => NodeDiameter * 1f;
+    private float MinXSegmentSpacing => NodeDiameter * 0.5f;
+    private float MaxXSegmentSpacing => NodeDiameter * 3f;
+
+    public XenoArtifactGraphControl()
+    {
+        IoCManager.InjectDependencies(this);
+        RobustXamlLoader.Load(this);
+
+        _font = new VectorFont(IoCManager.Resolve<IResourceCache>().GetResource<FontResource>("/EngineFonts/NotoSans/NotoSansMono-Regular.ttf"), 16);
+    }
+
+    public void SetArtifact(Entity<XenoArtifactComponent>? artifact)
+    {
+        _artifact = artifact;
+    }
+
+    protected override void KeyBindDown(GUIBoundKeyEventArgs args)
+    {
+        base.KeyBindDown(args);
+
+        if (args.Handled || args.Function != EngineKeyFunctions.UIClick)
+            return;
+
+        if (_hoveredNode == null)
+            return;
+        OnNodeSelected?.Invoke(_hoveredNode.Value);
+        UserInterfaceManager.ClickSound();
+    }
+
+    protected override void Draw(DrawingHandleScreen handle)
+    {
+        base.Draw(handle);
+
+        _hoveredNode = null;
+        if (_artifact == null)
+            return;
+
+        var cursor = (UserInterfaceManager.MousePositionScaled.Position * UIScale) - GlobalPixelPosition;
+
+        var artiSys = _entityManager.System<XenoArtifactSystem>();
+
+        var maxDepth = artiSys.GetAllNodes(_artifact.Value).Max(s => s.Comp.Depth);
```

Comment:
todo: replace XenoArtifactSystem calls with 1 access to complext cache-object, that contains all pre-calculated layers of data.

Reply:
client side contains all of net nodes, which means there is no state-hiding done currently. which means that we should not be afraid of node list sync-ing from backto front